File: python/src/docgen/clients/pdf_forms.py
"""PDF Forms sub-client: inspect and fill form fields."""

from __future__ import annotations

from typing import Any

from docgen._files import FileInput, to_base64
from docgen._serialization import from_dict, to_dict
from docgen._transport import Transport
from docgen.models.forms import PdfFormField, PdfFormFillRequest


class PdfFormsClient:
    """Client for PDF AcroForm operations."""

    def __init__(self, transport: Transport) -> None:
        self._transport = transport

    def inspect_fields(self, pdf: FileInput) -> list[PdfFormField]:
        """Inspect form fields in a PDF.

        Args:
            pdf: PDF file (path, bytes, or base64).

        Returns:
            List of form fields found in the PDF.
        """
        data = self._transport.request_json(
            "POST", "/api/pdf-forms/fields/base64",
            json={"pdfBase64": to_base64(pdf)},
        )
        # API returns a list directly
        if isinstance(data, list):
            return [from_dict(PdfFormField, item) for item in data]
        return [from_dict(PdfFormField, item) for item in data.get("fields", [])]

    def fill(
        self,
        pdf: FileInput,
        fields: dict[str, str],
        *,
        flatten: bool = False,
        password: str | None = None,
    ) -> bytes:
        """Fill form fields in a PDF.

        Args:
            pdf: PDF file with form fields.
            fields: Field values to fill (name → value).
            flatten: Make fields non-editable after filling.
            password: Optional password for the output PDF.

        Returns:
            Filled PDF bytes.
        """
        request = PdfFormFillRequest(
            pdf_base64=to_base64(pdf),
            fields=fields,
            flatten=flatten,
            password=password,
        )
        return self._transport.request_bytes(
            "POST", "/api/pdf-forms/fill", json=to_dict(request)
        )

File: typescript/src/clients/pdf-forms.ts
import type { Transport } from "../transport.js";
import type { PdfFormField } from "../models/forms.js";
import { toBase64, type FileInput } from "../files.js";

/** Client for PDF form inspection and filling. */
export class PdfFormsClient {
  constructor(private readonly transport: Transport) {}

  /** Inspect form fields in a PDF. */
  async inspectFields(file: FileInput): Promise<PdfFormField[]> {
    const base64 = toBase64(file);
    return this.transport.requestJson<PdfFormField[]>(
      "POST",
      "/api/pdf-forms/inspect",
      { pdfBase64: base64 },
    );
  }

  /** Fill form fields in a PDF. */
  async fill(
    file: FileInput,
    fields: Record<string, string>,
    flatten = false,
  ): Promise<Buffer> {
    const base64 = toBase64(file);
    return this.transport.requestBytes("POST", "/api/pdf-forms/fill", {
      pdfBase64: base64,
      fields,
      flatten,
    });
  }
}
